src/main/java/jenkins/plugins/publish_over_ssh/BapSshClient.java

```diff
@@ -198,18 +192,30 @@ public boolean makeDirectory(final String directory) {
     }
 
     public void transferFile(final BapSshTransfer bapSshTransfer, final FilePath filePath,
-                             final InputStream inputStream) throws SftpException {
-
-        if( !isAvoidSameFileUpload() || remoteResourceCache.checkCachedResource(filePath) ) {
-          buildInfo.printIfVerbose(Messages.console_put(filePath.getName()));
-          sftp.put(inputStream, filePath.getName());
-          success();
-        }
-        else
-        {
-          buildInfo.println(Messages._console_warning(
-            Messages.console_message_transferskip(filePath.getName()) )
-            .toString());
+                             final InputStream inputStream) throws SftpException, IOException, InterruptedException {
+        String fileName = filePath.getName();
+        buildInfo.printIfVerbose(Messages.console_put(fileName));
+        sftp.put(inputStream, fileName);
+
+        if (bapSshTransfer.isKeepFilePermissions()) {
+          FilePath parentFle = filePath.getParent();
+          if (parentFle != null) {
+            FilePath directory = parentFle.absolutize();
```

Comment:
Is it logically correct to change mode in this method? Maybe it'll be duplicated several times for one directory. Maybe there is some closer place for directory creation? `changeDirectory` or something else